crates/uv-auth/src/middleware.rs

```diff
@@ -985,6 +992,64 @@ mod tests {
         Ok(())
     }
 
+    #[test(tokio::test)]
+    async fn test_keyring_always_authenticate() -> Result<(), Error> {
+        let username = "user";
+        let password = "password";
+        let server = start_test_server(username, password).await;
+        let base_url = Url::parse(&server.uri())?;
+
+        let auth_policies = auth_policies_for(&base_url, AuthPolicy::Always);
+        let client = test_client_builder()
+            .with(
+                AuthMiddleware::new()
+                    .with_cache(CredentialsCache::new())
+                    .with_keyring(Some(KeyringProvider::dummy([(
+                        format!(
+                            "{}:{}",
+                            base_url.host_str().unwrap(),
+                            base_url.port().unwrap()
+                        ),
+                        username,
+                        password,
+                    )])))
+                    .with_url_auth_policies(auth_policies),
+            )
+            .build();
+
+        assert_eq!(
+            client.get(server.uri()).send().await?.status(),
+            200,
+            "Credentials (including a username) should be pulled from the keyring"
+        );
+
+        let mut url = base_url.clone();
+        url.set_username(username).unwrap();
+        assert_eq!(
+            client.get(url).send().await?.status(),
+            200,
+            "The password for the username should be pulled from the keyring"
+        );
+
+        let mut url = base_url.clone();
+        url.set_username(username).unwrap();
+        url.set_password(Some("invalid")).unwrap();
+        assert_eq!(
+            client.get(url).send().await?.status(),
+            401,
+            "Password in the URL should take precedence and fail"
+        );
+
+        let mut url = base_url.clone();
+        url.set_username("other_user").unwrap();
+        assert!(
+            matches!(client.get(url).send().await, Error::Middleware(_)),
```

Comment:
This error added in https://github.com/astral-sh/uv/pull/12313

It'd be nice to match on a specific type here.

Reply:
Agreed. In that case, we should also use variants [here](https://github.com/astral-sh/uv/blob/011a6de6dc1a49a4a6e816d8c5df39800a362879/crates/uv-auth/src/middleware.rs#L354) and [here](https://github.com/astral-sh/uv/blob/011a6de6dc1a49a4a6e816d8c5df39800a362879/crates/uv-auth/src/middleware.rs#L329).

Reply:
The problem is that this is in the context of the `reqwest_middleware::Middleware` trait. So `handle()` needs to return a `reqwest_middleware::Result<Response>`.

Reply:
Ah. It might not be worth it then. Thanks for looking anyway.